packages/pluggableWidgets/rich-text-web/src/components/RichText.tsx

```diff
@@ -0,0 +1,111 @@
+    advancedContentFilter?: { allowedContent: string; disallowedContent: string } | null;
+    onValueChange?: (value: string) => void;
+    onKeyChange?: () => void;
+    onKeyPress?: () => void;
+}
+
+export const RichTextEditor = (props: RichTextProps): ReactElement => {
+    const {
+        editorType,
+        plugins,
+        enterMode,
+        shiftEnterMode,
+        value,
+        readOnly,
+        readOnlyStyle,
+        advancedContentFilter
+    } = props;
+    const [element, setElement] = useState<HTMLElement | null>(null);
+    const { width, height } = props.dimensions
+        ? getDimensions({ ...props.dimensions })
+        : {
+              width: "100%",
+              height: "100%"
+          };
+    const [ckeditorConfig, setCkeditorConfig] = useState<CKEditorHookProps<any>>({
+        element,
+        editorUrl: "/widgets/ckeditor/ckeditor.js",
+        type: editorType,
+        config: {
+            autoGrow_minHeight: 300,
+            toolbarCanCollapse: true,
+            autoGrow_onStartup: true,
+            width,
+            height,
+            enterMode: defineEnterMode(enterMode || ""),
+            shiftEnterMode: defineEnterMode(shiftEnterMode || ""),
+            disableNativeSpellChecker: !props.spellChecker
+        },
+        initContent: value,
+        dispatchEvent: ({ type, payload }) => {
+            if (type === CKEditorEventAction.change) {
+                const value = payload.editor.getData() as string;
+                if (props.onKeyChange) {
+                    props.onKeyChange();
+                }
+                if (props.onKeyPress) {
+                    props.onKeyPress();
+                }
+                if (props?.onValueChange) {
+                    const content = props.sanitizeContent ? sanitizeHtml(value) : value;
+                    props?.onValueChange(content);
+                }
+            }
+        },
+        subscribeTo: ["change"]
+    });
+    const key = useMemo(() => Date.now(), [ckeditorConfig]);
+    useEffect(() => {
+        const config = { ...props.toolbar };
+        if (plugins?.length) {
+            plugins.forEach(plugin => addPlugin(plugin, config));
+        }
+        if (advancedContentFilter) {
+            config.allowedContent = advancedContentFilter.allowedContent;
+            config.disallowedContent = advancedContentFilter.disallowedContent;
+        }
+
+        setCkeditorConfig({
+            ...ckeditorConfig,
+            initContent: value,
+            element,
+            config: {
+                ...ckeditorConfig.config,
+                ...config
+            }
+        });
+    }, [props]);
+    return (
+        <div
+            className={classNames(props.class, "widget-rich-text", `${readOnly ? `editor-${readOnlyStyle}` : ""}`)}
+            style={{ width, height }}
+        >
+            <div ref={setElement} id={props.name} />
+            <MainEditor config={ckeditorConfig} key={key} />
```

Comment:
I don't quite get how the actual editor is actually rendered onto the screen. We store the above div element in state and set some CKEditor specific configurations. In `MainEditor`, we use the CKEditor hook, but it doesn't return or render anything as far as I see. So what is responsible for putting the CKEditor on the screen?

Reply:
In `MainEditor` component it destroys `CKEditor` instance and creates a new one (``` useCKEditor(config);```)

here, useCKEditor generates ckeditor instance every time when some of the props are changed.

It's looking for div `(<div ref={setElement} id={props.name} />)` and placing it with an `element` property which is passed through `ckeditorConfig` object

[element  - Reference to underlying to DOM element.](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_react_v2.html#api-2)

Why we need to destroy and initialise ckeditor again and again? Because it doesn't provide api to update config object dynamically.

`stringAttribute` is a dynamic property and it has status (pending, completed) or something like that, (maybe you can explain why it's working in this way), but here the problem was, when you set `readOnly` config `false` on your editor props, u always receive true at the beginning and then false at the end. And because there is no option to dynamically change readOnly value on CKEditor you need to destroy and initialise again.

I hope I answered sufficiently

Reply:
So https://github.com/mendix/widgets-resources/pull/1125/files#diff-85d3cc8ae664c556ca3ca667615281345e689ec33861bdd6ff56cc051b1ae809R92 is responsible for rendering the CKEditor onto the `element`?

Reply:
Yes, if you look at the initial definition you will see that I am passing `element` property there. Element is a ref element from div